### ISSUE TITLE
Policy plugin failure in daemons

### ISSUE THREAD

**Issue by astroclark** [rucio/rucio#5117]
I would like to use a policy plugin, mostly for the purposes of deploying our own LFN2PFN algorithm, but with a view to supporting our (LIGO/IGWN) own schema and permissions in future.

I have managed to deploy [this](https://git.ligo.org/rucio/gravcat-policy) package server containers using the 1.26.10 release as a base image and all looks fine with the following helm configs:

```
  policy:
    package: "gravcat_policy"
    permission: "permission"
    schema: "schema"
    lfn2pfn_algorithm_default: "igwn_frames"
```

However, when I attempt to deploy the daemons, all k8s pods crash with traceback:
```
$  kubectl logs rucio-daemons-abacus-collection-replica-6f446969cc-gpg2v
rucio.cfg not found. will generate one.
INFO:root:Merged 74 configuration values from /tmp/rucio.config.default.cfg
INFO:root:Merged 42 configuration values from /opt/rucio/etc/rucio.config.common.json
INFO:root:Merged 0 configuration values from /opt/rucio/etc/rucio.config.component.json
INFO:root:Merged 42 configuration values from ENV
starting daemon with: abacus-collection-replica --threads 2

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/rucio/common/schema/__init__.py", line 63, in <module>
    module = importlib.import_module(POLICY)
  File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 941, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.6/site-packages/gravcat_policy/__init__.py", line 26, in <module>
    from rucio.rse.protocols.protocol import RSEDeterministicTranslation
  File "/usr/local/lib/python3.6/site-packages/rucio/rse/__init__.py", line 104, in <module>
    from rucio.core.rse import get_rse_protocols, get_rse_id
ImportError: cannot import name 'get_rse_protocols'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/rucio-abacus-collection-replica", line 30, in <module>
    from rucio.daemons.abacus.collection_replica import run, stop
  File "/usr/local/lib/python3.6/site-packages/rucio/daemons/abacus/collection_replica.py", line 35, in <module>
    import rucio.db.sqla.util
  File "/usr/local/lib/python3.6/site-packages/rucio/db/sqla/util.py", line 48, in <module>
    from rucio.core.account_counter import create_counters_for_new_account
  File "/usr/local/lib/python3.6/site-packages/rucio/core/account_counter.py", line 29, in <module>
    import rucio.core.account
  File "/usr/local/lib/python3.6/site-packages/rucio/core/account.py", line 41, in <module>
    import rucio.core.rse
  File "/usr/local/lib/python3.6/site-packages/rucio/core/rse.py", line 60, in <module>
    from rucio.core.rse_counter import add_counter, get_counter
  File "/usr/local/lib/python3.6/site-packages/rucio/core/rse_counter.py", line 27, in <module>
    from rucio.db.sqla import models, filter_thread_work
  File "/usr/local/lib/python3.6/site-packages/rucio/db/sqla/models.py", line 49, in <module>
    from rucio.common.schema import get_schema_value
  File "/usr/local/lib/python3.6/site-packages/rucio/common/schema/__init__.py", line 65, in <module>
    raise exception.PolicyPackageNotFound('Module ' + POLICY + ' not found')
rucio.common.exception.PolicyPackageNotFound: The specified policy package cannot be loaded
Details: Module gravcat_policy.schema not found
```
Note that I am *only* specifying the configuration in my values.yaml; I do not currently mount my own `rucio.cfig` or `rucio.config.common.json`.

Finally, I note that a manual module import works fine if i leave policy settings at default values.  i.e.,:
```
$  kubectl exec rucio-daemons-abacus-account-75754d9d44-6zxzh -- python3 -c "import importlib; importlib.import_module('gravcat_policy.schema')"
```

**Comment by bari12** [comment]
Hi James,
This error is new to me and it's not clear why it works on the server, but not on the daemon. I wonder if there is some kind of circular dependency issue here 
```
...
File "/usr/local/lib/python3.6/site-packages/rucio/rse/__init__.py", line 104, in <module>
    from rucio.core.rse import get_rse_protocols, get_rse_id
ImportError: cannot import name 'get_rse_protocols'
```
Seems definitely weird.

@jamesp-epcc any idea?

**Comment by astroclark** [comment]
TL;DR: importing `RSEDeterministicTranslation` in the policy package `__init__.py` breaks things.

I gave up on 1.26.10 and just spent a couple of days digging around with 1.27.   On 1.27.2 on the *server*, i ultimately see:
```
[root@a041f4d20b12 tmp]# python3 -c "import rucio.core.config as config_db"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/rucio/core/config.py", line 31, in <module>
    from rucio.db.sqla import models
  File "/usr/local/lib/python3.6/site-packages/rucio/db/sqla/models.py", line 50, in <module>
    from rucio.common.schema import get_schema_value
  File "/usr/local/lib/python3.6/site-packages/rucio/common/schema/__init__.py", line 63, in <module>
    module = importlib.import_module(POLICY)
  File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/usr/local/lib/python3.6/site-packages/gravcat_policy/__init__.py", line 26, in <module>
    from rucio.rse.protocols.protocol import RSEDeterministicTranslation
  File "/usr/local/lib/python3.6/site-packages/rucio/rse/__init__.py", line 95, in <module>
    from rucio.core.rse import get_rse_protocols, get_rse_id
  File "/usr/local/lib/python3.6/site-packages/rucio/core/rse.py", line 56, in <module>
    import rucio.core.account_counter
  File "/usr/local/lib/python3.6/site-packages/rucio/core/account_counter.py", line 29, in <module>
    import rucio.core.account
  File "/usr/local/lib/python3.6/site-packages/rucio/core/account.py", line 45, in <module>
    from rucio.core.vo import vo_exists
  File "/usr/local/lib/python3.6/site-packages/rucio/core/vo.py", line 26, in <module>
    import rucio.core.config as config_db
AttributeError: module 'rucio.core' has no attribute 'config'
```
After much poking, I think you're right about the circular dependency.  The culprit seems to be the:
```
from rucio.rse.protocols.protocol import RSEDeterministicTranslation
```
in my policy package `__init__.py` (as suggested in [1](https://rucio.readthedocs.io/en/latest/policy_packages.html))

Moving that import into the lfn2pfn file instead seems to eliminate this issue - I'm now working on getting the config to recognise the desired lfn2pfn algorithm.

**Comment by astroclark** [comment]
Ok, resolved this all by:
1. *Removing* `RSEDeterministicTranslation` import from the policy package `__init__.py` 
2.  Add the `RSEDeterministicTranslation` import and lfn2pfn registration to the lfn2pfn module itself
3. Configured the server and daemons with e.g.:
```
  policy:
    package: "gravcat_policy"
    lfn2pfn_module: "gravcat_policy.lfn2pfn"
    lfn2pfn_algorithm_default: "igwn_frames"
```
If this is the intended behavior, suggest updating these [docs](https://rucio.readthedocs.io/en/latest/policy_packages.html), which do say to do that import at the `__init__.py` (unless I'm misreading).

I'm comfortable closing this ticket since I have a solution but I'll leave it open in case it would be helpful to discuss/test any fixes to the `__init__.py` level import of `RSEDeterministicTranslation`.

**Comment by bari12** [comment]
To me the workaround seems reasonable, but it should indeed be documented. Pinging @jamesp-epcc again on this, maybe he has an opinion.

**Comment by jamesp-epcc** [comment]
Apologies, due to my GitHub notification settings being misconfigured I have only just seen this. I will take a look. I did run into circular dependency issues when testing the policy packages so it could well be one of those.

**Comment by jamesp-epcc** [comment]
Actually, after a quick look at this I think this is exactly the same issue I ran into last year. When I initially developed the policy packages, I thought that the packages could import `RSEDeterministicTranslation` and then call `register` on it to register their own LFN2PFN algorithms, however this doesn't work in practice due to the import loop problem.

The new way to register LFN2PFN algorithms is for the policy package to implement a method that returns a dictionary of them in its `__init__.py`. Here is how DUNE's one looks:

```
def get_lfn2pfn_algorithms():
    return { 'DUNE': lfn2pfn_DUNE }
```

(This may change again if my latest PR is merged to make the algorithm registration more generic, in which case the code will look like this:

```
def get_algorithms():
    return { 'lfn2pfn': { 'DUNE': lfn2pfn_DUNE }, 'surl': { 'DUNE': construct_surl_dune } }
```
)

If this doesn't resolve the problem, please let me know and I will look into it further.
